R/qCount.R

```diff
@@ -3,516 +3,895 @@
         tid <- seq_along(bh) - 1L
-        if(!is.null(targets)) {
-            if(any(is.na(i <- match(targets,names(bh)))))
-                stop(sprintf("some targets not found in bamfile '%s': %s",bamfile,targets[which(is.na(i))]))
+        if (!is.null(targets)) {
+            if (any(is.na(i <- match(targets, names(bh)))))
+                stop(sprintf("some targets not found in bamfile '%s': %s",
+                             bamfile, targets[which(is.na(i))]))
             bh <- bh[i]
             tid <- tid[i]
         }
-        start <- rep(0L, length(tid)) # samtool library has 0-based inclusiv start
+        start <- rep(0L, length(tid)) # samtools library has 0-based inclusive start
         end <- unname(bh) ## samtool library has 0-based exclusiv end
         # count junctions
-        count <- .Call(countJunctions, bamfile, tid, start, end, allelic, includeSecondary, mapqmin, mapqmax)
+        count <- .Call(countJunctions, bamfile, tid, start, end, allelic,
+                       includeSecondary, mapqmin, mapqmax)
         return(count)
 
     }, error = function(ex) {
-        emsg <- paste("Internal error on", Sys.info()['nodename'], "query bamfile", bamfile, 
+        emsg <- paste("Internal error on", Sys.info()['nodename'], 
+                      "query bamfile", bamfile, 
                       "\n Error message is:", ex$message)
         stop(emsg)
     })
-    
 }
 
 
 ## count alignments (with the C-function) for single bamfile, single shift, and single set of regions
 ## return a numeric vector with length(regions) elements (same order as regions)
+#' @keywords internal
+#' @importFrom Rsamtools scanBamHeader
+#' @importFrom GenomicRanges seqnames
+#' @importFrom BiocGenerics strand start end match as.vector
 countAlignments <- function(bamfile, regions, shift, selectReadPosition, orientation,
                             useRead, broaden, allelic, includeSpliced, includeSecondary,
-                            mapqmin, mapqmax, absisizemin, absisizemax)
-{
+                            mapqmin, mapqmax, absisizemin, absisizemax) {
     tryCatch({ # try catch block goes through the whole function
         
         ## translate seqnames to tid and create region data.frame
-        seqnamesBamHeader <- names(scanBamHeader(bamfile)[[1]]$targets)
+        seqnamesBamHeader <- names(Rsamtools::scanBamHeader(bamfile)[[1]]$targets)
         
         ## prepare region vectors
         #tid <- IRanges::as.vector(IRanges::match(seqnames(regions), seqnamesBamHeader)) - 1L
-        tid <- as.vector(match(seqnames(regions), seqnamesBamHeader) - 1L) 
-        start <- start(regions) - 1L ## samtool library has 0-based inclusiv start
-        end <- end(regions) ## samtool library has 0-based exclusiv end
+        tid <- BiocGenerics::as.vector(BiocGenerics::match(GenomicRanges::seqnames(regions),
+                                                           seqnamesBamHeader) - 1L) 
```

Comment:
This is a good example relating to the above comments: we have an `@importFrom BiocGenerics as.vector`, so we are importing the generic. Now here we are calling the generic on a `integer-Rle` that really needs the method `S4Vectors::as.vector.Rle`. Is it found only by change because we already are loading the `S4Vectors` namespace anyway? As a thought experiment: If this would be the only place in `QuasR` where we need something from `S4Vectors`, would it suffice to `@importFrom BiocGenerics as.vector`?

Reply:
I have retained the current code here since it seems to do the right thing (and `S4Vectors` only exports an S3 method `as.vector`).